pylabnet/scripts/pulsemaster/pulsemaster.py

```diff
@@ -0,0 +1,1457 @@
+        or lists of the same length.
+
+        Generate list out of dictionary.
+        """
+
+        values = datadict.values()
+        data_ok = False
+
+        # Check if all values are one allowed datatype:
+        allowed_datatypes = [str, int, float]
+        if all([type(value) in allowed_datatypes for value in values]):
+                data_ok = True
+                datadict = self._prepare_single_string_dict(datadict)
+                return data_ok, datadict
+
+        # Check if values are all lists.
+        if all(isinstance(value, list) for value in values):
+
+            # Check if lists are of same length.
+            it = iter(values)
+            the_len = len(next(it))
+
+            if not all(len(l) == the_len for l in it):
+                data_ok = False
+                return data_ok, None
+            else:
+                data_ok = True
+                datadict = self._prepare_list_dict(datadict)
+                return data_ok, datadict
+        else:
+            data_ok = False
+            return data_ok, None
+
+    def data(self, index, role):
+        """ From https://stackoverflow.com/questions/28186118/how-to-make-qtableview-to-enter-the-editing-mode-only-on-double-click"""
+        if not index.isValid(): return QVariant()
+        row=index.row()
+        column=index.column()
+
+        if row>len(self.datadict): return QVariant()
+        if column>len(self.datadict[row]): return QVariant()
+
+        if role == Qt.EditRole or role == Qt.DisplayRole:
+            return QVariant(self.datadict[row][column])
+
+        return QVariant()
+
+    def setData(self, index, value, role=Qt.EditRole):
+        """ From https://stackoverflow.com/questions/28186118/how-to-make-qtableview-to-enter-the-editing-mode-only-on-double-click"""
+        if index.isValid():
+            if role == Qt.EditRole:
+                row = index.row()
+                column=index.column()
+                if row>len(self.datadict) or column>len(self.datadict[row]):
+                    return False
+                else:
+                    self.datadict[row][column]=value
+                    self.dataChanged.emit(index, index)
+                    return True
+        return False
+
+    def rowCount(self, index):
+        # The length of the outer list.
+        return len(self.datadict)
+
+    def columnCount(self, index):
+        # The following takes the first sub-list, and returns
+        # the length (only works if all rows are an equal length)
+        return len(self.datadict[0])
+
+
+    def add_dict(self, data_dict):
+        """Populate table from dictionary.
+
+        :data_dict: Dictionary to populate table from.
+        """
+        for i, (key, item) in enumerate(data_dict.items()):
+            key = QTableWidgetItem(key)
+            item = QTableWidgetItem(item)
+            self.setItem(i,0,key)
+            self.setItem(i,1,str(item))
+
+
+class AddPulseblockPopup(QWidget):
```

Comment:
in parallel i did implement something similar in what i thought was a sort of quick and dirty way - see `pylabnet.scripts.data_center.datasets.ParameterPopup` in the g2 branch. Would be interested in your thoughts - should we have a generic parameter prompting popup in the `external_gui.Window` module?

Reply:
I agree, we should factor out commonly used GUI parts. I've created an issue for this (#144).